components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/UserAdmin.java

```diff
@@ -280,19 +281,26 @@ private void addUserRole(String roleName, String[] userList, String[] permission
         if (!isAllowedRoleName(roleName, realm)) {
             throw new UserAdminException("Role name is reserved by the system.");
         }
-        if (!isInternalRole) {
-            if (getUserAdminProxy().isRoleAndGroupSeparationEnabled()) {
-                if (ArrayUtils.isNotEmpty(permissions)) {
-                    addInternalSystemRole(roleName, permissions);
-                    permissions = new String[0];
-                }
-            }
+        if (isInternalRole) {
+            getUserAdminProxy().addInternalRole(roleName, userList, permissions);
+        } else if (!(getUserAdminProxy().isRoleAndGroupSeparationEnabled()) || ArrayUtils.isEmpty(permissions)) {
             getUserAdminProxy().addRole(roleName, userList, permissions, isSharedRole);
         } else {
-            getUserAdminProxy().addInternalRole(roleName, userList, permissions);
+            String internalSystemRoleName = getInternalSystemRoleName(roleName);
+            getUserAdminProxy().addInternalRole(internalSystemRoleName, new String[0], permissions);
+            permissions = new String[0];
```

Comment:
why we are removing permissions here

Reply:
Because when we adding the role (actually treated as group) we shouldn't add permissions but we have to add the permissions to internal roles. So line number 290 need permissions but line number 292 doesn't need permission